incubator/virtualcluster/pkg/syncer/resources/namespace/checker.go

```diff
@@ -76,98 +79,90 @@ func (c *controller) shouldBeGabageCollected(ns *v1.Namespace) bool {
+				Key:          conversion.ToSuperMasterNamespace(cluster, vList.Items[i].GetName()),
+			})
 		}
+	}
 
-		if shouldDelete {
-			opts := &metav1.DeleteOptions{
-				PropagationPolicy: &constants.DefaultDeletionPolicy,
-				Preconditions:     metav1.NewUIDPreconditions(string(pNamespace.UID)),
-			}
-			if err := c.namespaceClient.Namespaces().Delete(context.TODO(), pNamespace.Name, *opts); err != nil {
-				klog.Errorf("error deleting pNamespace %s in super master: %v", pNamespace.Name, err)
-			} else {
-				metrics.CheckerRemedyStats.WithLabelValues("DeletedOrphanSuperMasterNamespaces").Inc()
-			}
+	d := differ.HandlerFuncs{}
+	d.AddFunc = func(vObj differ.ClusterObject) {
+		if err := c.MultiClusterController.RequeueObject(vObj.OwnerCluster, vObj.Object); err != nil {
+			klog.Errorf("error requeue vNamespace %v in cluster %s: %v", vObj.GetName(), vObj.GetOwnerCluster(), err)
+		} else {
+			metrics.CheckerRemedyStats.WithLabelValues("RequeuedTenantNamespaces").Inc()
 		}
 	}
-}
+	d.UpdateFunc = func(vObj, pObj differ.ClusterObject) {
+		v := vObj.Object.(*v1.Namespace)
+		p := pObj.Object.(*v1.Namespace)
 
-// checkNamespacesOfTenantCluster checks to see if namespaces in specific cluster keeps consistency.
-func (c *controller) checkNamespacesOfTenantCluster(clusterName string) {
-	listObj, err := c.MultiClusterController.List(clusterName)
-	if err != nil {
-		klog.Errorf("error listing namespaces from cluster %s informer cache: %v", clusterName, err)
-		return
+		// if vc object is deleted, we should reach here
+		if c.shouldBeGarbageCollected(p) || p.Annotations[constants.LabelUID] != string(v.UID) {
+			c.deleteNamespace(p)
+		}
 	}
-	klog.V(4).Infof("check namespaces consistency in cluster %s", clusterName)
-	namespaceList := listObj.(*v1.NamespaceList)
-	for i, vNamespace := range namespaceList.Items {
-		targetNamespace := conversion.ToSuperMasterNamespace(clusterName, vNamespace.Name)
-		pNamespace, err := c.nsLister.Get(targetNamespace)
-		if errors.IsNotFound(err) {
-			// pNamespace not found and vNamespace still exists, we need to create pNamespace again
-			if err := c.MultiClusterController.RequeueObject(clusterName, &namespaceList.Items[i]); err != nil {
-				klog.Errorf("error requeue vNamespace %s in cluster %s: %v", vNamespace.Name, clusterName, err)
-			} else {
-				metrics.CheckerRemedyStats.WithLabelValues("RequeuedTenantNamespaces").Inc()
+	d.DeleteFunc = func(pObj differ.ClusterObject) {
+		p := pObj.Object.(*v1.Namespace)
+		if p.Annotations[constants.LabelVCRootNS] == "true" {
+			if !c.shouldBeGarbageCollected(p) {
+				return
 			}
-			continue
 		}
+		c.deleteNamespace(p)
+	}
 
-		if err != nil {
-			klog.Errorf("error getting pNamespace %s from super master cache: %v", targetNamespace, err)
-		}
+	vSet.Difference(pSet, differ.FilteringHandler{
+		Handler: d,
+		FilterFunc: func(obj differ.ClusterObject) bool {
+			if obj.OwnerCluster == "" && obj.GetAnnotations()[constants.LabelVCRootNS] == "true" {
```

Comment:
Is it ok to remove the obj.OwnerCluster == "" condition?

Reply:
just skip vObj